### ISSUE TITLE
Build errors[E0061, E0277, E0308]

### ISSUE THREAD

**Issue by makorne** [RecordingTheSmile/ergoreq#3]
Hi! Thank you for your interesting crate!

cargo build

```
   ...
   Compiling reqwest v0.12.4
   Compiling ergoreq v0.1.1 
error[E0277]: the trait bound `http::error::Error: From<<reqwest::header::HeaderName as TryFrom<K>>::Error>` is not satisfied
   --> src/wrappers/request_builder_wrapper.rs:154:40
    |
154 |         self.inner = self.inner.header(key, value);
    |                                 ------ ^^^ the trait `From<<reqwest::header::HeaderName as TryFrom<K>>::Error>` is not implemented for `http::error::Error`, which is required by `<reqwest::header::HeaderName as TryFrom<K>>::Error: Into<http::error::Error>`
    |                                 |
    |                                 required by a bound introduced by this call
    |
    = help: the following other types implement trait `From<T>`:
              <http::error::Error as From<Infallible>>
              <http::error::Error as From<http::method::InvalidMethod>>
              <http::error::Error as From<http::status::InvalidStatusCode>>
              <http::error::Error as From<http::uri::InvalidUri>>
              <http::error::Error as From<http::uri::InvalidUriParts>>
              <http::error::Error as From<reqwest::header::InvalidHeaderName>>
              <http::error::Error as From<reqwest::header::InvalidHeaderValue>>
              <http::error::Error as From<reqwest::header::MaxSizeReached>>
    = note: required for `<reqwest::header::HeaderName as TryFrom<K>>::Error` to implement `Into<http::error::Error>`
note: required by a bound in `RequestBuilder::header`
   --> /home/dan/.cargo/registry/src/index.crates.io-6f17d22bba15001f/reqwest-0.12.4/src/async_impl/request.rs:192:44
    |
189 |     pub fn header<K, V>(self, key: K, value: V) -> RequestBuilder
    |            ------ required by a bound in this associated function
...
192 |         <HeaderName as TryFrom<K>>::Error: Into<http::Error>,
    |                                            ^^^^^^^^^^^^^^^^^ required by this bound in `RequestBuilder::header`

error[E0277]: the trait bound `http::error::Error: From<<reqwest::header::HeaderValue as TryFrom<V>>::Error>` is not satisfied
   --> src/wrappers/request_builder_wrapper.rs:154:45
    |
154 |         self.inner = self.inner.header(key, value);
    |                                 ------      ^^^^^ the trait `From<<reqwest::header::HeaderValue as TryFrom<V>>::Error>` is not implemented for `http::error::Error`, which is required by `<reqwest::header::HeaderValue as TryFrom<V>>::Error: Into<http::error::Error>`
    |                                 |
    |                                 required by a bound introduced by this call
    |
    = help: the following other types implement trait `From<T>`:
              <http::error::Error as From<Infallible>>
              <http::error::Error as From<http::method::InvalidMethod>>
              <http::error::Error as From<http::status::InvalidStatusCode>>
              <http::error::Error as From<http::uri::InvalidUri>>
              <http::error::Error as From<http::uri::InvalidUriParts>>
              <http::error::Error as From<reqwest::header::InvalidHeaderName>>
              <http::error::Error as From<reqwest::header::InvalidHeaderValue>>
              <http::error::Error as From<reqwest::header::MaxSizeReached>>
    = note: required for `<reqwest::header::HeaderValue as TryFrom<V>>::Error` to implement `Into<http::error::Error>`
note: required by a bound in `RequestBuilder::header`
   --> /home/dan/.cargo/registry/src/index.crates.io-6f17d22bba15001f/reqwest-0.12.4/src/async_impl/request.rs:194:45
    |
189 |     pub fn header<K, V>(self, key: K, value: V) -> RequestBuilder
    |            ------ required by a bound in this associated function
...
194 |         <HeaderValue as TryFrom<V>>::Error: Into<http::Error>,
    |                                             ^^^^^^^^^^^^^^^^^ required by this bound in `RequestBuilder::header`

error[E0308]: mismatched types
   --> src/wrappers/request_builder_wrapper.rs:201:41
    |
201 |         self.inner = self.inner.headers(headers);
    |                                 ------- ^^^^^^^ expected `reqwest::header::HeaderMap`, found `http::HeaderMap`
    |                                 |
    |                                 arguments to this method are incorrect
    |
    = note: `http::HeaderMap` and `reqwest::header::HeaderMap` have similar names, but are actually distinct types
note: `http::HeaderMap` is defined in crate `http`
   --> /home/dan/.cargo/registry/src/index.crates.io-6f17d22bba15001f/http-0.2.12/src/header/map.rs:45:1
    |
45  | pub struct HeaderMap<T = HeaderValue> {
    | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
note: `reqwest::header::HeaderMap` is defined in crate `http`
   --> /home/dan/.cargo/registry/src/index.crates.io-6f17d22bba15001f/http-1.1.0/src/header/map.rs:45:1
    |
45  | pub struct HeaderMap<T = HeaderValue> {
    | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
    = note: perhaps two different versions of crate `http` are being used?
note: method defined here
   --> /home/dan/.cargo/registry/src/index.crates.io-6f17d22bba15001f/reqwest-0.12.4/src/async_impl/request.rs:234:12
    |
234 |     pub fn headers(mut self, headers: crate::header::HeaderMap) -> RequestBuilder {
    |            ^^^^^^^

error[E0308]: mismatched types
   --> src/wrappers/request_builder_wrapper.rs:244:41
    |
244 |         self.inner = self.inner.version(version);
    |                                 ------- ^^^^^^^ expected `reqwest::Version`, found `http::Version`
    |                                 |
    |                                 arguments to this method are incorrect
    |
    = note: `http::Version` and `reqwest::Version` have similar names, but are actually distinct types
note: `http::Version` is defined in crate `http`
   --> /home/dan/.cargo/registry/src/index.crates.io-6f17d22bba15001f/http-0.2.12/src/version.rs:26:1
    |
26  | pub struct Version(Http);
    | ^^^^^^^^^^^^^^^^^^
note: `reqwest::Version` is defined in crate `http`
   --> /home/dan/.cargo/registry/src/index.crates.io-6f17d22bba15001f/http-1.1.0/src/version.rs:26:1
    |
26  | pub struct Version(Http);
    | ^^^^^^^^^^^^^^^^^^
    = note: perhaps two different versions of crate `http` are being used?
note: method defined here
   --> /home/dan/.cargo/registry/src/index.crates.io-6f17d22bba15001f/reqwest-0.12.4/src/async_impl/request.rs:372:12
    |
372 |     pub fn version(mut self, version: Version) -> RequestBuilder {
    |            ^^^^^^^

error[E0277]: the trait bound `http::HeaderName: reqwest::header::IntoHeaderName` is not satisfied
    --> src/wrappers/request_builder_wrapper.rs:285:32
     |
285  |                 headers.insert(http::header::COOKIE, cookie_header);
     |                         ------ ^^^^^^^^^^^^^^^^^^^^ the trait `reqwest::header::IntoHeaderName` is not implemented for `http::HeaderName`
     |                         |
     |                         required by a bound introduced by this call
     |
     = help: the following other types implement trait `reqwest::header::IntoHeaderName`:
               &'a reqwest::header::HeaderName
               &'static str
               reqwest::header::HeaderName
note: required by a bound in `reqwest::header::HeaderMap::<T>::insert`
    --> /home/dan/.cargo/registry/src/index.crates.io-6f17d22bba15001f/http-1.1.0/src/header/map.rs:1228:12
     |
1226 |     pub fn insert<K>(&mut self, key: K, val: T) -> Option<T>
     |            ------ required by a bound in this associated function
1227 |     where
1228 |         K: IntoHeaderName,
     |            ^^^^^^^^^^^^^^ required by this bound in `HeaderMap::<T>::insert`

error[E0277]: the trait bound `http::HeaderName: reqwest::header::IntoHeaderName` is not satisfied
    --> src/wrappers/request_builder_wrapper.rs:303:36
     |
303  |                     headers.insert(http::header::COOKIE, cookie_header);
     |                             ------ ^^^^^^^^^^^^^^^^^^^^ the trait `reqwest::header::IntoHeaderName` is not implemented for `http::HeaderName`
     |                             |
     |                             required by a bound introduced by this call
     |
     = help: the following other types implement trait `reqwest::header::IntoHeaderName`:
               &'a reqwest::header::HeaderName
               &'static str
               reqwest::header::HeaderName
note: required by a bound in `reqwest::header::HeaderMap::<T>::insert`
    --> /home/dan/.cargo/registry/src/index.crates.io-6f17d22bba15001f/http-1.1.0/src/header/map.rs:1228:12
     |
1226 |     pub fn insert<K>(&mut self, key: K, val: T) -> Option<T>
     |            ------ required by a bound in this associated function
1227 |     where
1228 |         K: IntoHeaderName,
     |            ^^^^^^^^^^^^^^ required by this bound in `HeaderMap::<T>::insert`

error[E0277]: the trait bound `http::HeaderName: reqwest::header::AsHeaderName` is not satisfied
   --> src/middleware/middleware.rs:49:26
    |
49  |                 .get_all(http::header::SET_COOKIE)
    |                  ------- ^^^^^^^^^^^^^^^^^^^^^^^^ the trait `reqwest::header::AsHeaderName` is not implemented for `http::HeaderName`
    |                  |
    |                  required by a bound introduced by this call
    |
    = help: the following other types implement trait `reqwest::header::AsHeaderName`:
              &'a reqwest::header::HeaderName
              &'a std::string::String
              &'a str
              reqwest::header::HeaderName
              std::string::String
note: required by a bound in `reqwest::header::HeaderMap::<T>::get_all`
   --> /home/dan/.cargo/registry/src/index.crates.io-6f17d22bba15001f/http-1.1.0/src/header/map.rs:828:12
    |
826 |     pub fn get_all<K>(&self, key: K) -> GetAll<'_, T>
    |            ------- required by a bound in this associated function
827 |     where
828 |         K: AsHeaderName,
    |            ^^^^^^^^^^^^ required by this bound in `HeaderMap::<T>::get_all`

error[E0277]: the trait bound `http::HeaderName: reqwest::header::IntoHeaderName` is not satisfied
    --> src/middleware/middleware.rs:64:29
     |
64   |                     .insert(http::header::COOKIE, header_value);
     |                      ------ ^^^^^^^^^^^^^^^^^^^^ the trait `reqwest::header::IntoHeaderName` is not implemented for `http::HeaderName`
     |                      |
     |                      required by a bound introduced by this call
     |
     = help: the following other types implement trait `reqwest::header::IntoHeaderName`:
               &'a reqwest::header::HeaderName
               &'static str
               reqwest::header::HeaderName
note: required by a bound in `reqwest::header::HeaderMap::<T>::insert`
    --> /home/dan/.cargo/registry/src/index.crates.io-6f17d22bba15001f/http-1.1.0/src/header/map.rs:1228:12
     |
1226 |     pub fn insert<K>(&mut self, key: K, val: T) -> Option<T>
     |            ------ required by a bound in this associated function
1227 |     where
1228 |         K: IntoHeaderName,
     |            ^^^^^^^^^^^^^^ required by this bound in `HeaderMap::<T>::insert`

error[E0308]: mismatched types
    --> src/middleware/middleware.rs:64:51
     |
64   |                     .insert(http::header::COOKIE, header_value);
     |                      ------                       ^^^^^^^^^^^^ expected `HeaderValue`, found `http::HeaderValue`
     |                      |
     |                      arguments to this method are incorrect
     |
     = note: `http::HeaderValue` and `HeaderValue` have similar names, but are actually distinct types
note: `http::HeaderValue` is defined in crate `http`
    --> /home/dan/.cargo/registry/src/index.crates.io-6f17d22bba15001f/http-0.2.12/src/header/value.rs:21:1
     |
21   | pub struct HeaderValue {
     | ^^^^^^^^^^^^^^^^^^^^^^
note: `HeaderValue` is defined in crate `http`
    --> /home/dan/.cargo/registry/src/index.crates.io-6f17d22bba15001f/http-1.1.0/src/header/value.rs:22:1
     |
22   | pub struct HeaderValue {
     | ^^^^^^^^^^^^^^^^^^^^^^
     = note: perhaps two different versions of crate `http` are being used?
help: the return type of this call is `http::HeaderValue` due to the type of the argument passed
    --> src/middleware/middleware.rs:62:17
     |
62   | /                 request
63   | |                     .headers_mut()
64   | |                     .insert(http::header::COOKIE, header_value);
     | |___________________________________________________------------^
     |                                                     |
     |                                                     this argument influences the return type of `insert`
note: method defined here
    --> /home/dan/.cargo/registry/src/index.crates.io-6f17d22bba15001f/http-1.1.0/src/header/map.rs:1226:12
     |
1226 |     pub fn insert<K>(&mut self, key: K, val: T) -> Option<T>
     |            ^^^^^^

error[E0277]: the trait bound `http::HeaderName: reqwest::header::AsHeaderName` is not satisfied
   --> src/middleware/auto_redirect_middleware.rs:49:22
    |
49  |                 .get(http::header::LOCATION)
    |                  --- ^^^^^^^^^^^^^^^^^^^^^^ the trait `reqwest::header::AsHeaderName` is not implemented for `http::HeaderName`
    |                  |
    |                  required by a bound introduced by this call
    |
    = help: the following other types implement trait `reqwest::header::AsHeaderName`:
              &'a reqwest::header::HeaderName
              &'a std::string::String
              &'a str
              reqwest::header::HeaderName
              std::string::String
note: required by a bound in `reqwest::header::HeaderMap::<T>::get`
   --> /home/dan/.cargo/registry/src/index.crates.io-6f17d22bba15001f/http-1.1.0/src/header/map.rs:753:12
    |
751 |     pub fn get<K>(&self, key: K) -> Option<&T>
    |            --- required by a bound in this associated function
752 |     where
753 |         K: AsHeaderName,
    |            ^^^^^^^^^^^^ required by this bound in `HeaderMap::<T>::get`

error[E0308]: `match` arms have incompatible types
  --> src/middleware/auto_redirect_middleware.rs:73:22
   |
71 |               let new_method = match response.status().as_u16() {
   |  ______________________________-
72 | |                 307 | 308 => origin_method.to_owned(),
   | |                              ------------------------ this is found to be of type `reqwest::Method`
73 | |                 _ => Method::GET,
   | |                      ^^^^^^^^^^^ expected `reqwest::Method`, found `http::Method`
74 | |             };
   | |_____________- `match` arms have incompatible types
   |
   = note: `http::Method` and `reqwest::Method` have similar names, but are actually distinct types
note: `http::Method` is defined in crate `http`
  --> /home/dan/.cargo/registry/src/index.crates.io-6f17d22bba15001f/http-0.2.12/src/method.rs:46:1
   |
46 | pub struct Method(Inner);
   | ^^^^^^^^^^^^^^^^^
note: `reqwest::Method` is defined in crate `http`
  --> /home/dan/.cargo/registry/src/index.crates.io-6f17d22bba15001f/http-1.1.0/src/method.rs:45:1
   |
45 | pub struct Method(Inner);
   | ^^^^^^^^^^^^^^^^^
   = note: perhaps two different versions of crate `http` are being used?

error[E0308]: mismatched types
  --> src/middleware/auto_redirect_middleware.rs:85:42
   |
85 |             *new_request.headers_mut() = origin_headers.to_owned();
   |             --------------------------   ^^^^^^^^^^^^^^^^^^^^^^^^^ expected `http::HeaderMap`, found `reqwest::header::HeaderMap`
   |             |
   |             expected due to the type of this binding
   |
   = note: `reqwest::header::HeaderMap` and `http::HeaderMap` have similar names, but are actually distinct types
note: `reqwest::header::HeaderMap` is defined in crate `http`
  --> /home/dan/.cargo/registry/src/index.crates.io-6f17d22bba15001f/http-1.1.0/src/header/map.rs:45:1
   |
45 | pub struct HeaderMap<T = HeaderValue> {
   | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
note: `http::HeaderMap` is defined in crate `http`
  --> /home/dan/.cargo/registry/src/index.crates.io-6f17d22bba15001f/http-0.2.12/src/header/map.rs:45:1
   |
45 | pub struct HeaderMap<T = HeaderValue> {
   | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
   = note: perhaps two different versions of crate `http` are being used?

error[E0277]: the trait bound `reqwest::Request: TryFrom<http::Request<Vec<u8>>>` is not satisfied
  --> src/middleware/auto_redirect_middleware.rs:87:57
   |
87 |             response = inner_client.execute(new_request.try_into()?).await?;
   |                                                         ^^^^^^^^ the trait `From<http::Request<Vec<u8>>>` is not implemented for `reqwest::Request`, which is required by `http::Request<Vec<u8>>: TryInto<_>`
   |
   = note: required for `http::Request<Vec<u8>>` to implement `Into<reqwest::Request>`
   = note: required for `reqwest::Request` to implement `TryFrom<http::Request<Vec<u8>>>`
   = note: required for `http::Request<Vec<u8>>` to implement `TryInto<reqwest::Request>`

error[E0277]: `?` couldn't convert the error to `error::Error`
  --> src/middleware/auto_redirect_middleware.rs:87:67
   |
87 |             response = inner_client.execute(new_request.try_into()?).await?;
   |                                                                   ^ the trait `From<Infallible>` is not implemented for `error::Error`, which is required by `std::result::Result<reqwest::Response, error::Error>: FromResidual<std::result::Result<Infallible, _>>`
   |
   = note: the question mark operation (`?`) implicitly performs a conversion on the error value using the `From` trait
   = help: the following other types implement trait `From<T>`:
             <error::Error as From<OutOfRangeError>>
             <error::Error as From<anyhow::Error>>
             <error::Error as From<http::Error>>
             <error::Error as From<reqwest::Error>>
   = note: required for `std::result::Result<reqwest::Response, error::Error>` to implement `FromResidual<std::result::Result<Infallible, Infallible>>`

error[E0061]: this method takes 2 arguments but 1 argument was supplied
  --> src/middleware/auto_retry_middleware.rs:44:30
   |
44 |                 match self.0.should_retry(current_retry_times) {
   |                              ^^^^^^^^^^^^ ------------------- an argument of type `SystemTime` is missing
   |
note: method defined here
  --> /home/dan/.cargo/registry/src/index.crates.io-6f17d22bba15001f/retry-policies-0.4.0/src/retry_policy.rs:6:8
   |
6  |     fn should_retry(&self, request_start_time: SystemTime, n_past_retries: u32) -> RetryDecision;
   |        ^^^^^^^^^^^^
help: provide the argument
   |
44 |                 match self.0.should_retry(/* SystemTime */, current_retry_times) {
   |                                          ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

error[E0277]: cannot subtract `chrono::DateTime<Utc>` from `SystemTime`
  --> src/middleware/auto_retry_middleware.rs:46:61
   |
46 |                         let should_wait_for = execute_after - Utc::now();
   |                                                             ^ no implementation for `SystemTime - chrono::DateTime<Utc>`
   |
   = help: the trait `Sub<chrono::DateTime<Utc>>` is not implemented for `SystemTime`
   = help: the following other types implement trait `Sub<Rhs>`:
             <SystemTime as Sub<OffsetDateTime>>
             <SystemTime as Sub<cookie::time::Duration>>
             <SystemTime as Sub<std::time::Duration>>

Some errors have detailed explanations: E0061, E0277, E0308.
For more information about an error, try `rustc --explain E0061`.
error: could not compile `ergoreq` (lib) due to 16 previous errors
```
cargo 1.79.0 (ffa9cf99a 2024-06-03)

**Comment by RecordingTheSmile** [comment]
Thanks for reporting this issue! It seems that reqwest 0.12 has changed the `http` crate version, caused this happens. A new version will be published to solve this problem very soon.

**Comment by RecordingTheSmile** [comment]
This issue has been resolved in the latest version 0.2.0 related to #4 . You can upgrade to this version to resolve this issue. If you find that the latest version still has other problems, you are welcome to create an issue for discussion!
